src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -348,6 +406,61 @@ private void EnsureSignature()
             }
         }
 
+        /// <summary>
+        /// If the backing field is unknown, set it to null.
+        /// </summary>
+        /// <remarks>
+        /// This should be called only if we're sure the backing field can't be non-null value.
+        /// </remarks>
+        internal void MarkBackingFieldAsCalculated()
+        {
+            Interlocked.CompareExchange(ref _lazyBackingFieldSymbol, null, _lazyBackingFieldSymbolSentinel);
```

Comment:
> Can we assert that ```_lazyBackFieldSymbol``` is null after this, to ensure that some other thread didn't set a different value?

I don''t think this expectation is justified.

Reply:
> I don't think this expectation is justified.

How so? I would expect that all threads should arrive at the same conclusion, or we have a race condition.

Reply:
> How so? I would expect that all threads should arrive at the same conclusion, or we have a race condition.

All threads definitely should arrive to the same conclusion. However, that is not always null.

Reply:
If we're in `MarkBackingFieldAsCalculated`, why could it be anything but `null`?

Reply:
> If we're in `MarkBackingFieldAsCalculated`, why could it be anything but `null`?

Because we might have already calculated it to be something else. That is the whole point of doing ```CompareExchange```. When we call ```MarkBackingFieldAsCalculated``` we know that we have done all the work that could assign non-null value.

Reply:
We either have a "known" value by this point, or it will definitely be null if we force calculate it.

Reply:
Indeed, as @AlekseyTs noted, the assertion will fail. Per doc comment, `**If** the backing field is unknown, set it to null`. I think the confusion may have came from the `remarks` comment, I'll adjust it.

Reply:
> We either have a "known" value by this point, or it will definitely be null if we force calculate it.

I think this is where my confusion is coming from: given the method name and implementation, I was expecting that this would only ever be called when `_lazyBackingFieldSymbol` was unset, not that it was a backstop of sorts.